Modules.md

```diff
@@ -0,0 +1,160 @@
+
+```bnf
+extend global_decl :
+  | module_decl_main
+;
+
+template_elaborated_ident :
+  (module_path '::')? ident _disambiguate_template template_list ?
+```
+
+## Behaviour-changing items
+
+These items cannot be imported, but they affect behavior:
+
+- [Extensions](https://www.w3.org/TR/WGSL/#extensions)
+- [Global diagnostic filters](https://www.w3.org/TR/WGSL/#global-diagnostic-directive)
+
+These are always set at the very top in the main module, and affect all loaded files. They come before the imports.
+
+When, during parsing of loaded files and modules, we encounter usage of a module that has an extension or a global diagnostic filter within it, we check if the main module enables it, or sets the filter.
+
+If yes, everything is fine. If not, we throw an error.
+
+## Bundling WESL Files
+
+As modules are imported into the global namespace, a naive bundling algorithm would be:
+- Remove all `load` statements
+- Concatenate files together according to the linking logic described in [Imports](./Imports.md) 
+
+## Linking WESL Files
+
+Linking WESL files with modules is a little bit more complex than the base [Imports](./Imports.md) linking algorithm.
+
+To do so the following steps must be taken:
+- Bundling needs to occur prior to linking. 
+- A map of modules to their implementation should be made. A canonical module name should be picked using the logic described in the Name Mangling section in this proposal. 
+- Module usages should be replaced with the canonical module name.
+- Next symbols within the module need to be mangled using the canonical module name using the [mangling algorithm](./NameMangling.md) and should be added to the global scope. 
+- Usage of symbols within a module should then replaced with the mangled representation using the same algorithm. 
+- Finally a plain wgsl module is produced
+
+### Name Mangling
+Symbols within a module are mangled based on the module path, using the scheme described in [Name Mangling](./NameMangling.md). 
+
+As modules can be aliased, the module path choosen by the mangling scheme uses the following algorithm:  
+ - Picks the path with the fewest number of parts in it
+    - If this is a tie, picks the path string with the smallest number of unicode characters in it
+        - Finally if that is also a tie, uses a region invariant string comparision function, and chooses the path that is alphabetically first (inclusive of the `::` operators)
+
+This scheme means that if a given symbol has been included into the global scope, it is not mangled at all. It has the absolute minimum number of parts in the path!
+
+An example of how to apply the algorithm described above, would be the name used in mangling the `quat_from_euler` function. The name chosen to be mangled would be `Math::Float::quat_from_euler`, rather than `Math::FloatMath::quat_from_euler`. This is because while they have the same number of parts in their paths, `Math::Float::quat_from_euler` has fewer number of characters.
```

Comment:
Hmmm. Interesting point. My motivation was originally to reduce the size of the code using this approach. Most modules outside of the global scope would contain implementation details and not be for public use anyway. 

However bindings are a bit of an exception as users may still want to be able to introspect on them and to also generate binding code. 

So a dilemma!

Reply:
Regarding "generate binding code", would it be reasonable to say that that is part of what a good linker should offer? Or rather, a good linker should offer some language specific APIs for it.

As in, we aren't expecting people to generate perfect bindings from generated code. Generated code could, for example, no longer have comments. Good bindings code would include documentation comments.

So for the generated output, we only need something that is predictable and usable. (compared to raw WGSL)

Reply:
That makes sense. Do you have a suggestion for achieving this? Would prefer if we didn't have to special case the output names of entry points though understand it might be necessary.